test/replication-luatest/gh_6033_box_promote_demote_test.lua

```diff
@@ -0,0 +1,325 @@
+end
+
+local make_follower = function(server)
+    server:exec(function()
+        box.ctl.demote()
+    end)
+    server:wait_election_state('follower')
+end
+
+g.before_each(function(g)
+    g.cluster = cluster:new({})
+
+    g.box_cfg = {
+        election_mode = 'off';
+        replication_timeout = 0.1,
+        replication = {
+            helpers.instance_uri('server_', 1);
+            helpers.instance_uri('server_', 2);
+            helpers.instance_uri('server_', 3);
+        };
+    }
+
+    g.server_1 = g.cluster:build_and_add_server(
+        {alias = 'server_1', box_cfg = g.box_cfg})
+    g.server_2 = g.cluster:build_and_add_server(
+        {alias = 'server_2', box_cfg = g.box_cfg})
+    g.server_3 = g.cluster:build_and_add_server(
+        {alias = 'server_3', box_cfg = g.box_cfg})
+    g.cluster:start()
+
+    g.server_3:exec(function()
+        box.ctl.promote()
+    end)
+    g.server_3:wait_synchro_queue_owner()
+    g.server_1:wait_synchro_queue_owner(g.server_3:instance_id())
+    g.server_2:wait_synchro_queue_owner(g.server_3:instance_id())
+end)
+
+g.after_each(function(g)
+    g.cluster:stop()
+    g.cluster.servers = nil
+    g.cluster = nil
+end)
+
+g.before_test('test_leader_promote', function(g)
+    g.server_1:box_config({election_mode = 'manual'})
+    make_leader(g.server_1)
+end)
+
+g.before_test('test_raft_leader_promote', function(g)
+    g.server_1:box_config({election_mode = 'manual'})
+end)
+
+g.before_test('test_follower_demote', function(g)
+    g.server_1:box_config({election_mode = 'manual'})
+    make_follower(g.server_1)
+end)
+
+g.before_test('test_voter_promote', function(g)
+    g.server_1:box_config({election_mode = 'voter'})
+end)
+
+g.before_test('test_fail_limbo_acked_promote', function(g)
+    g.server_1:box_config({
+        replication_synchro_quorum = 2,
+        replication_timeout = 0.5,
+        replication = {
+            helpers.instance_uri('server_', 1);
+            helpers.instance_uri('server_', 2);
+        },
+    })
+    g.server_2:box_config({
+        read_only = true,
+        replication_synchro_quorum = 2,
+        replication_timeout = 0.5,
+        replication = {
+            helpers.instance_uri('server_', 1);
+            helpers.instance_uri('server_', 2);
+        },
+    })
+end)
+
+g.before_test('test_raft_leader_demote', function(g)
+    g.server_1:box_config({election_mode = 'manual'})
+    make_leader(g.server_1)
+end)
+
+-- Test that box_promote and box_demote
+-- will return 0 if server is not configured.
+g.test_unconfigured = function()
+    local ok, _ = pcall(box.ctl.promote)
+    luatest.assert(ok, 'error while promoting unconfigured server')
+
+    local ok, _ = pcall(box.ctl.demote)
+    luatest.assert(ok, 'error while demoting unconfigured server')
+end
+
+-- Test that box_promote will return 0
+-- if server is already raft leader and limbo owner.
+g.test_leader_promote = function(g)
+    local ok, _ = g.server_1:exec(function()
```

Comment:
So, here in the test code it is more obvious - in the comment you said "server is already raft leader" but from the test code it is hardly visible. Because the first promotion is done somewhere above in a `before_test` function. Please, move the preparations into the test cases themselves.